train/ngram.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+public class Ngram
+{
+    public void ngram_n2(string sentence, int N)
+    {
+        string[] Sentence = sentence.Split(" ");
+        //split to single word
+        List<string> completed = new List<string>();
+        //save the result
+        if (sentence == null)
+        {
+            Console.WriteLine("The test sentence is empty");
+        }
+        else
+        {
+            for (int i = 0; i <= (Sentence.Length - N); i++)
+            //loop word list
+            {
+                List<string> temp = new List<string>();
+                for (int j = i; j < i + N; j++)
+                //complete one n-gram each loop
+                {
+                    temp.Add(Sentence[j]);
+                }
+                completed.Add(string.Join(" ", temp));
+            }
+            for (int i = 0; i < completed.Count; i++)
+            {
+                Console.WriteLine(completed[i]);
+            }
+        }
+    }
+    public IEnumerable<string> ngram_queue(string sentence, int N)
+    {
+        sentence = sentence + " ";
+        StringBuilder completed = new StringBuilder();
+        //save the result
+        Queue<int> Word_len = new Queue<int>();
+        //save the length of each word
+        int count = 0;
+        //count the word have been catched
+        int lastWord = 0;
+        //the length of lastword in queue
+        if (sentence == null)
+        {
+            Console.WriteLine("The test sentence is empty");
+        }
+        for (int i = 0; i < sentence.Length; i++)
+        {
+            if (char.IsLetterOrDigit(sentence[i]) || sentence[i] == '\'')
+            //make sure the sentence[i] is letter digit or '
+            {
+                completed.Append(sentence[i]);
+                lastWord++;
+            }
+            else
+            //when suffer the " "(a word completed)
+            {
+                if (lastWord > 0)
+                //if there are letters in buffer
+                {
+                    Word_len.Enqueue(lastWord);
+                    lastWord = 0;
+                    count++;
+                    if (count == N)
+                    //if the counted words equal to requied N
+                    {
+                        yield return completed.ToString();
+                        completed.Remove(0, Word_len.Dequeue() + 1);
+                        count -= 1;
+                    }
+                    completed.Append(" ");
+                }
+            }
+        }
+    }
+
+    public void readNgram(string sentence, int N)
+    {
+        Console.WriteLine("1-gram");
+        foreach (int i in Enumerable.Range(0, sentence.Split(" ").Count()))
```

Comment:
I think for this iteration, it could be better if you write it like:

var sentencePieces = sentence.Split(' ');
for (int i = 0; i < sentencePieces.Length; i++)
{
  Console.WriteLine(sentencePieces[i]);
}

Two reasons here:
1. Get rid of creating your sentence array (by split) during each iteration. Instead, just create it at the beginning, so that you will get less code executions.
2. It might not be appropriate to use Enumerable.Range here to iterate. As you already split sentence into array, calling "Length" property of an array instance (O(1)) seems much better than creating a sequence.

Reply:
Here I just want to practice Enumerable.Range()  function. Predeclare the sentencePiece is efficient. I will update my code

Reply:
Here I just want to practice Enumerable.Range() function. Predeclaring the sentencePiece is efficient. I will update my code.